### PR TITLE
PPSD Fix exact trace cutting; includes test

### LINKED ISSUE #2039: Minor correction in obspy.signal.PPSD
Hi, was trying to do very precise PSD processing of our archive. Unfortunately I need to work with exact samples and expect 48 segments (30 min. each) per day. The PPSD calculation always uses a single sample per segment more than it should and results in to the final daily segment being dropped.

Here is some demo code:

    import obspy
    import numpy as np

    header = {
        "starttime":  obspy.UTCDateTime("2017-01-01T00:00:00"),
        "network": "GB",
        "station": "MCH1",
        "channel": "BHE"
    }

    # 24.5 hours to account for 50% overlap through the next day
    print tr = obspy.Trace(data=np.arange(30 * 60 * 49, dtype=np.int32), header=header)
    >>> GB.MCH1..BHE | 2017-01-01T00:00:00.000000Z - 2017-01-02T00:29:59.000000Z | 1.0 Hz, 88200 samples

    ppsd = obspy.signal.PPSD(
        tr.stats,
        obspy.read_inventory("http://www.orfeus-eu.org/fdsnws/station/1/query?sta=MCH1&cha=BHE&level=response"),
    )

    ppsd.add(tr)

    print len(ppsd._times_processed)
    >>> 47

When we add a single second extra to the trace it will give 48 PSDs as expected but that is incorrect.

Here is a list of the segments it calculates:

    GB.MCH1..BHE | 2017-01-01T00:00:00.000000Z - 2017-01-01T01:00:00.000000Z | 1.0 Hz, 3601 samples
    GB.MCH1..BHE | 2017-01-01T00:30:00.000000Z - 2017-01-01T01:30:00.000000Z | 1.0 Hz, 3601 samples
    ...
    GB.MCH1..BHE | 2017-01-01T22:30:00.000000Z - 2017-01-01T23:30:00.000000Z | 1.0 Hz, 3601 samples
    GB.MCH1..BHE | 2017-01-01T23:00:00.000000Z - 2017-01-02T00:00:00.000000Z | 1.0 Hz, 3601 samples

It shows that the final sample used per segment overlaps and we are using 3601 samples per segment, not 3600 as we should. IMHO It can be fixed easily by subtracting tr.stats.delta from the end time of each segment.

I think it could be changed, what do you reckon @megies?

### PR DESCRIPTION
**Pull request by Jollyfant**
Fixes #2039

### CODE REVIEW

**Comment by megies** [comment]
Would be cool if you added a test that checks the length of the time series going into the individual PSD calculations.

**Comment by Jollyfant** [comment]
Is there a private keeping track of this..? I can't find it.

**Comment by megies** [comment]
Probably not.. 

Actually, there's a property that keeps the expected number of samples in a slice though..

```python
 348         # Trace length for one psd segment.
 349         self._len = int(self.sampling_rate * self.ppsd_length)
 350         # make an initial dummy psd and to get the array of periods
 351         _, freq = mlab.psd(np.ones(self.len), self.nfft,
 352                            self.sampling_rate, noverlap=self.nlap)
```

..and it's also checked when the individual PSDs are calculated:

```python
 746     def __process(self, tr):
 747         """
 748         Processes a segment of data and save the psd information.
 749         Whether `Trace` is compatible (station, channel, ...) has to
 750         checked beforehand.
 751        
 752         :type tr: :class:`~obspy.core.trace.Trace`
 753         :param tr: Compatible Trace with data of one PPSD segment
 754         :returns: `True` if segment was successfully processed,
 755             `False` otherwise.
 756         """
 757         # XXX DIRTY HACK!!
 758         if len(tr) == self.len + 1:
 759             tr.data = tr.data[:-1]
 760         # one last check..
 761         if len(tr) != self.len:
 762             msg = "Got a piece of data with wrong length. Skipping"
 763             warnings.warn(msg)
 764             print(len(tr), self.len)
 765             return False
```

..and there's also some code that was probably used to handle such 1-sample-too-long slices..

**Comment by Jollyfant** [comment]
So we can check if `ppsd.len` is equal 3600 in my test case? That should be OK.

**Comment by megies** [comment]
> So we can check if ppsd.len is equal 3600 in my test case? That should be OK. (fyi it is 👍)

Ok, so the problem is reduced to later slices (e.g. at the end of daily stream object) having slightly later starttime (by a few samples) due to the malformed slicing.. and thus the last slice being excluded since it's starting some milliseconds later then expected? Nice catch!

**Comment by Jollyfant** [comment]
Yeah that is basically just it, nothing worrisome.

**Comment by Jollyfant** [comment]
Can this go in maintenance_1.1.x since this is essentially a bug fix? Or maybe that's later than 1.1.0.. idk!

File: obspy/signal/tests/test_spectral_estimation.py [hunk 1]
```diff
@@ -677,6 +677,32 @@ def test_ppsd_temporal_plot(self):
...
+
+        header = {
+            "starttime":  UTCDateTime("2017-01-01T00:00:00"),
+            "network": "GR",
+            "station": "FUR",
+            "channel": "BHZ"
+        }
+
+        # 49 segments of 30 minutes to allow 30 minutes overlap in next day
+        tr = Trace(data=np.arange(30 * 60 * 49, dtype=np.int32), header=header)
```
Comment (megies) [inline obspy/signal/tests/test_spectral_estimation.py]: not sure how long this test runs.. but maybe using a smaller number of slices (3? 4?) is equally fit as a regression test? the problem should show right away after a single slice, I guess?

**Comment by megies** [comment]
Personally, I'm OK if this goes into maintenance.

File: obspy/signal/tests/test_spectral_estimation.py (hunk 1 above)
Comment (Jollyfant) [reply obspy/signal/tests/test_spectral_estimation.py]: Good point! I will do three segments.

**Comment by Jollyfant** [comment]
What going on with CI? It's bailing on unrelated floating point accuracies.

**Comment by krischer** [comment]
> What going on with CI? It's bailing on unrelated floating point accuracies.

Oh this is nasty :( They changed the string representation of arrays with numpy 1.14.0.

**Comment by krischer** [comment]
After looking at it in a bit more detail its maybe not that bad: The printing style can be chosen and has support for a "legacy" mode. I'll open a new PR and then we can rebase the others on it.

**Comment by Jollyfant** [comment]
> After looking at it in a bit more detail its maybe not that bad: The printing style can be chosen and has support for a "legacy" mode. I'll open a new PR and then we can rebase the others on it.

Can we stick that in one central place or do all scripts need updating?

**Comment by krischer** [comment]
I'll push later today - it took a bit longer because there are also a ton of new warnings in numpy 1.14. I put it in the central `obspy-runtests` script and also added configuration files for pytest so it works for those without a lot of effort.

**Comment by krischer** [comment]
Done in #2044.

**Comment by krischer** [comment]
The failures should be gone now if you rebase.

**Comment by Jollyfant** [comment]
Okay! Rebased and force pushed with a changelog entry. We pick `maintenance_1.1.x` or `master`?

**Comment by krischer** [comment]
I would tend to do this in `master` as it does change the calculations (however slightly) but I'll leave it up to @megies .

**Comment by Jollyfant** [comment]
Nah master is fine if we go for biannual releases 🐇

File: obspy/signal/tests/test_spectral_estimation.py [hunk 2]
```diff
@@ -677,6 +677,32 @@ def test_ppsd_temporal_plot(self):
...
+        )
+
+        ppsd.add(tr)
+
+        self.assertEqual(3, len(ppsd._times_processed))
+        self.assertEqual(3600, ppsd.len)
+
+        for i, time in enumerate(ppsd._times_processed):
+            current = UTCDateTime("2017-01-01T00:00:00") + (i * 30 * 60)
+            self.assertTrue(UTCDateTime(time) == current)
```
Comment (megies) [inline obspy/signal/tests/test_spectral_estimation.py]: interesting.. so I guess start times of individual PSD slices were never tested so far.. :-/

**Pull request merged**